File: 2021/python/day_13.py
def part_1():
    dots, fold_instructions = set(), []
    with open("inputs/day13", "r") as file:
        for line in file.read().splitlines():
            if line.startswith("fold along"):
                fold = line.split("=")
                fold_instructions.append((fold[0][-1], int(fold[1])))
            elif line != "":
                dots.add(tuple(map(int, line.split(","))))

    axis, position = fold_instructions[0]
    if axis == "x":
        dots = set(
            (position - abs(x - position), y) if x > position else (x, y)
            for (x, y) in dots
        )
    elif axis == "y":
        dots = set(
            (x, position - abs(y - position)) if y > position else (x, y)
            for (x, y) in dots
        )

    return len(dots)


def part_2():
    dots, fold_instructions = set(), []
    with open("inputs/day13", "r") as file:
        for line in file.read().splitlines():
            if line.startswith("fold along"):
                fold = line.split("=")
                fold_instructions.append((fold[0][-1], int(fold[1])))
            elif line != "":
                dots.add(tuple(map(int, line.split(","))))

    for axis, position in fold_instructions:
        if axis == "x":
            dots = set(
                (position - abs(x - position), y) if x > position else (x, y)
                for (x, y) in dots
            )
        elif axis == "y":
            dots = set(
                (x, position - abs(y - position)) if y > position else (x, y)
                for (x, y) in dots
            )

    max_x = max(x for (x, y) in dots)
    max_y = max(y for (x, y) in dots)
    for y in range(max_y + 1):
        for x in range(max_x + 1):
            print("#" if (x, y) in dots else " ", end="")
        print()
    return len(dots)


print(part_1())
print(part_2())

File: 2021/src/day_13.rs
pub const TITLE: &str = "--- Day 13: Transparent Origami ---";

use std::{collections::HashSet, fs};

fn folder(axis: char, position: i64, dots: HashSet<(i64, i64)>) -> HashSet<(i64, i64)> {
    let mut update: HashSet<(i64, i64)> = HashSet::with_capacity(dots.len());
    match axis {
        'x' => {
            for (x, y) in dots.iter() {
                if *x > position {
                    update.insert((2 * position - *x, *y));
                } else {
                    update.insert((*x, *y));
                }
            }
        }
        'y' => {
            for (x, y) in dots.iter() {
                if *y > position {
                    update.insert((*x, 2 * position - *y));
                } else {
                    update.insert((*x, *y));
                }
            }
        }
        _ => panic!("Invalid axis!"),
    }

    update
}

pub fn part_1() -> i64 {
    let mut once = false;

    fs::read_to_string("inputs/day13")
        .expect("Unable to read input")
        .lines()
        .filter(|line| !line.is_empty())
        .map(|s| s.to_string())
        .fold(HashSet::new(), |mut dots, line| {
            match line.starts_with("fold along") {
                false => {
                    let mut point = line.split(",");
                    let x = point.next().unwrap().parse::<i64>().unwrap();
                    let y = point.next().unwrap().parse::<i64>().unwrap();
                    dots.insert((x, y));

                    dots
                }
                true => match once {
                    false => {
                        once = true;
                        let mut point = line.split("=");
                        let axis = point.next().unwrap().chars().last().unwrap();
                        let position = point.next().unwrap().parse::<i64>().unwrap();

                        folder(axis, position, dots)
                    }
                    true => dots,
                },
            }
        })
        .len() as i64
}

pub fn part_2() -> i64 {
    fs::read_to_string("inputs/day13")
        .expect("Unable to read input")
        .lines()
        .filter(|line| !line.is_empty())
        .map(|s| s.to_string())
        .fold(HashSet::new(), |mut dots, line| {
            match line.starts_with("fold along") {
                false => {
                    let mut point = line.split(",");
                    let x = point.next().unwrap().parse::<i64>().unwrap();
                    let y = point.next().unwrap().parse::<i64>().unwrap();
                    dots.insert((x, y));

                    dots
                }
                true => {
                    let mut point = line.split("=");
                    let axis = point.next().unwrap().chars().last().unwrap();
                    let position = point.next().unwrap().parse::<i64>().unwrap();

                    folder(axis, position, dots)
                }
            }
        })
        .len() as i64
}

#[test]
fn test_part_1() {
    assert_eq!(part_1(), 775);
}

#[test]
fn test_part_2() {
    assert_eq!(part_2(), 102);
}
